### PR TITLE
add momentum control and sparse check for Adagrad, RMSprop, AdaDelta,…

### PR DESCRIPTION
**Pull request by backyes**
E.G, 

```
  285 void decayedAdagradApply(BaseMatrix& value,
  286                          BaseMatrix& grad,
  287                          BaseMatrix& mom,
  288                          BaseMatrix& accum,
  289                          BaseMatrix& lr,
  290                          real accumulatedRou,
  291                          real rou,
  292                          real epsilon,
  293                          real learningRate,
  294                          real momentum,
  295                          real decayRate,
  296                          bool firstTime) {
  297   // E(g_t^2) = \rou * E(g_{t-1}^2) + (1-\rou) * g^2
  298   // For the first time update, make the sum be the current square
  299   // so that the initial estimation of E(g_t^2) will not be too small.
  300   if (firstTime) {
  301     accum = accumulatedRou * accum + grad.square();
  302   } else {
  303     accum = accumulatedRou * accum + ((real)1 - rou) * grad.square();
  304   }
  305
  306   // learn_rate = 1/sqrt( ( E(g_t^2) + epsilon )
  307   // Basiclly if the bigger the magnitude gradient is,
  308   // the smaller the learning rate will be.
  309   lr = (accum + epsilon).sqrt().reciprocal();
  310
  311   mom = mom * momentum - learningRate * lr * (grad + value * decayRate);
  312   value += mom;
  313 }

```

The above implementation for DecayedAdagradOptimizer support adding additional momentum property. 

Adagrad, RMSprop, AdaDelta also support momentum control. 

In reality, if sparse update was set for any parameters, the momentum control is not allowed, because even the gradient of some weight was ZERO, the momentum of that is not ZERO, so the value of this weight should be updated to abey momentum rules.

### CODE REVIEW

**Comment by backyes** [comment]
还未完成，需要测试，先让大家看看思路是否正确。

File: python/paddle/trainer_config_helpers/optimizers.py [hunk 1]
```diff
@@ -246,18 +268,29 @@ class DecayedAdaGradOptimizer(BaseSGDOptimizer):
...
     :type epsilon: float
+    param momentum: add momentum for DecayedAdaGradOptimizer
+                    DecayedAdaGradOptimizer support add momentum property
+    type momentum: float
+    param sparse: if using sparse updater, set it to True
+    type sparse: bool
     """
 
+    def extra_settings(self):
+        default_momentum(self.momentum)
```
Comment (pengli09) [inline python/paddle/trainer_config_helpers/optimizers.py]: 这里不应该改default_momentum吧，default_momentum会修改全局默认值，如果self.momentum和用户手工设置的default_momentum不一致，将会给用户造成非常大的困扰。强烈建议不要在用户不明确知晓的情况下，修改全局默认值。

**Review by pengli09 — CHANGES_REQUESTED** [review]

File: python/paddle/trainer_config_helpers/optimizers.py (hunk 1 above)
Comment (backyes) [inline python/paddle/trainer_config_helpers/optimizers.py]: 这是参照MomentumOptimizer实现来的。

另外这个momemum设置是在optimizer上设置的，翻译成全局momentum也是合理的，同时用户仍然可以在parameter的attribute属性中设置momentum来覆盖paraemter level 的momentum。 

所以，我觉得还是合理的，@pengli09  觉得呢？

Comment (pengli09) [inline python/paddle/trainer_config_helpers/optimizers.py]: 那我觉着至少要在文档里加一处明显的提示吧，提醒用户会修改全局的默认值。如果没有提示的话，对于从老config格式切换过来的用户来讲很有可能引起意外的麻烦。另外是不是可以在default_momentum（其他的设置全局默认值的函数里也类似）中加一个检查，如果发现修改前的值与paddle自带的全局默认值不同的时候——也就是全局默认值被修改了一次以上，意味着有可能使用上出问题了——在log里打印出一条warning信息出来。这样即使出现了问题，也比较容易追查到源头。

Comment (backyes) [inline python/paddle/trainer_config_helpers/optimizers.py]: 如电话沟通： 

* 结合新接口封装逻辑，从momentum算法的momentum参数设置来看，新接口更加倾向于不调用config_parser.py的全局函数 * def default_momentum(val): * ， 而是在优化算法里设置momentum。 但是，由于新接口下，用户仍然可以在使用*新接口*情况下，调用config_parser.py的底层全局函数，这一定程度上容易造成困惑。  拿现在这个patch中的例子来说， 如果遵循新接口不调用底层全局函数来说，那么adagrad算法配合momentum也应通过在adagrad算法的属性中来添加momentum属性，达到和momentum的联动， 但是*粗心的用户可能仍然调用全局函数设置momentum，所以造成最终的mometnum含义不明确*。 

请 @reyoung  review 『新接口不建议调用config_parse.py 的全局函数』的结论，是否靠谱？ 

所以， 暂时综合考虑： 

*  将momentum和adagrad等优化算法的结合，在新接口下，抽象为adagrad算法的一个属性， *并在python注释中强调*。 这个思路也符合底层代码实现。
* 对每个优化算法添加sparse 属性（参照已有的momentum优化算法示例），使能check sparse和momentum属性的互斥性。（除了专门设计的sparse momentum算法，其他所有算法如何设置了momentum，理论上都不支持sparse训练，否则不符合理论）
* 为了新接口中，逐渐过渡并摒弃对底层config_parse.py全局函数的调用，添加对全局函数的调用监控，如果新接口配置中又调用了老接口config_parser.py函数，那么报deprecated warning。 （理论上，用户仍然可以跳过新接口，使用老接口）

**Comment by qingqing01** [comment]
虽然Paddle里是AdaGrad、AdaDelt、Adam等梯度更新，都加了momentum的变量。 但是paper里一般这些公式并不包含momentum，其他torch等平台，这些优化算法的更新也不包含momentum。设置这些参数的同时又设置momentum感觉并不常见吧？ 而且是否合理？

**Comment by pengli09** [comment]
@qingqing01 momentum+adagrad之类这种组会确实是会有人用的，之前在paper里也见过，而且本质上它们也不矛盾。

**Comment by backyes** [comment]
@pengli09 @reyoung 

对全局函数调用做了增强，且做了若干异常配置的test，通过。 


@reyoung 但是标准的config的单测部分依赖protostr，由于更新了单测代码，正确的更新protostr的姿势是? （我手动执行生成protostr的脚本不能解决问题）

**Comment by backyes** [comment]
因为很多单元测试 读取配置文件， 多次调用全局python模型配置函数，为了持这种实现，略去严格的abort异常检测，使用warning对潜在的异常配置进行logging，辅助诊断。

@pengli09

File: python/paddle/trainer/config_parser.py [hunk 2]
```diff
@@ -172,6 +172,35 @@ def config_func(func):
...
+    Some global function should be called only once in model config. For
+    history reason, global momentum is set with default_momentum, but
+    momentum also can be set as property of optimizer in
+    trainer_config_helper as well. This decorator tries to fix it.
+    Discussion: https://github.com/PaddlePaddle/Paddle/pull/1129
+    """
+
+    assert not hasattr(func, "is_called"), \
+        func.func_name + " already has is_called attribute"
+    setattr(func, "is_called", False)
```
Comment (reyoung) [inline python/paddle/trainer/config_parser.py]: 这里不需要有184-186行。

File: python/paddle/trainer/config_parser.py [hunk 3]
```diff
@@ -172,6 +172,35 @@ def config_func(func):
...
+
+    def wrapper(*args, **kargs):
+        if func.is_called:
```
Comment (reyoung) [inline python/paddle/trainer/config_parser.py]: 这里实现成

```python
if hasattr(func, 'is_called'):
    logger.warning(...)
setattr(func, 'is_called', True)
```

File: python/paddle/trainer/config_parser.py [hunk 4]
```diff
@@ -172,6 +172,35 @@ def config_func(func):
...
+    history reason, global momentum is set with default_momentum, but
+    momentum also can be set as property of optimizer in
+    trainer_config_helper as well. This decorator tries to fix it.
+    Discussion: https://github.com/PaddlePaddle/Paddle/pull/1129
+    """
+
+    assert not hasattr(func, "is_called"), \
+        func.func_name + " already has is_called attribute"
+    setattr(func, "is_called", False)
+
```
Comment (reyoung) [inline python/paddle/trainer/config_parser.py]: 使用[functools.wraps](https://docs.python.org/2/library/functools.html#functools.wraps)，否则docstring会丢掉的。

File: python/paddle/trainer/config_parser.py [hunk 5]
```diff
@@ -172,6 +172,35 @@ def config_func(func):
     return func
 
 
+def call_only_once(func):
+    """
+    Some global function should be called only once in model config. For
```
Comment (reyoung) [inline python/paddle/trainer/config_parser.py]: 注释改成
```python
"""
call_only_once is used to print warnings if the function is invoked multiple times.
"""
```

注释说 XXX是什么。 而不是 XXX为什么要这么做

为什么要这么做可以写到git message里面

File: python/paddle/trainer/config_parser.py [hunk 6]
```diff
@@ -172,6 +172,35 @@ def config_func(func):
...
+            logger.warning("executing " + func.func_name +
+                           " more than one time")
+        setattr(func, "is_called", True)
+
+        return func(*args, **kargs)
+
+    # rename wrapper to func.func_name for @config_func, otherwise the
+    # function decorated by call_only_once can not be used in model
+    # config file.
+    wrapper.func_name = func.func_name
```
Comment (reyoung) [inline python/paddle/trainer/config_parser.py]: 去掉这行，用functools.wraps

File: python/paddle/trainer_config_helpers/optimizers.py [hunk 7]
```diff
@@ -218,18 +229,29 @@ class RMSPropOptimizer(BaseSGDOptimizer):
...
             'learning_method': 'rmsprop',
             'ada_rou': self.rho,
             'ada_epsilon': self.epsilon
         }
 
-    def __init__(self, rho=0.95, epsilon=1e-6):
+    def __init__(self, rho=0.95, epsilon=1e-6, momentum=None, sparse=False):
         self.rho = rho
         self.epsilon = epsilon
+        assert not (sparse and momentum is not None)
```
Comment (reyoung) [inline python/paddle/trainer_config_helpers/optimizers.py]: 这个sparse是在哪用的？似乎，不应该在这里面做check，因为用户也不一定会按照道理传sparse参数进来。

File: python/paddle/trainer_config_helpers/optimizers.py [hunk 8]
```diff
@@ -218,18 +229,29 @@ class RMSPropOptimizer(BaseSGDOptimizer):
...
+        self.momentum = momentum
+        self.sparse = sparse
```
Comment (reyoung) [inline python/paddle/trainer_config_helpers/optimizers.py]: 这个sparse没用到啊，没必要保存下来吧？

File: python/paddle/trainer_config_helpers/optimizers.py [hunk 9]
```diff
@@ -246,18 +268,29 @@ class DecayedAdaGradOptimizer(BaseSGDOptimizer):
...
             'learning_method': 'decayed_adagrad',
             'ada_rou': self.rho,
             'ada_epsilon': self.epsilon
         }
 
-    def __init__(self, rho=0.95, epsilon=1e-6):
+    def __init__(self, rho=0.95, epsilon=1e-6, momentum=None, sparse=False):
         self.rho = rho
         self.epsilon = epsilon
+        assert not (sparse and momentum is not None)
```
Comment (reyoung) [inline python/paddle/trainer_config_helpers/optimizers.py]: 同上，没必要在这里check，回头我们再找个地方check这玩意。

File: python/paddle/trainer_config_helpers/optimizers.py [hunk 10]
```diff
@@ -277,18 +310,29 @@ class AdaDeltaOptimizer(BaseSGDOptimizer):
...
             'learning_method': 'adadelta',
             'ada_rou': self.rho,
             'ada_epsilon': self.epsilon
         }
 
-    def __init__(self, rho=0.95, epsilon=1e-6):
+    def __init__(self, rho=0.95, epsilon=1e-6, momentum=None, sparse=False):
         self.rho = rho
         self.epsilon = epsilon
+        assert not (sparse and momentum is not None)
```
Comment (reyoung) [inline python/paddle/trainer_config_helpers/optimizers.py]: 同上

File: python/paddle/trainer_config_helpers/tests/configs/test_fc.py [hunk 11]
```diff
@@ -1,6 +1,27 @@
...
+
+2. default_momentum and momentum attribution should be set exclusively
+   for better understanding.
+default_momentum(0.4)
+settings(batch_size=1000,
+         learning_method=AdaGradOptimizer(momentum=0.9, sparse=True),
+         learning_rate=1e-5)
+
+In theory any optimizer with momentum setting do not work with sparse upate
+except MomentumOptimizer which has implemented specical algorithm.
```
Comment (reyoung) [inline python/paddle/trainer_config_helpers/tests/configs/test_fc.py]: image

请用grammerly或者word检查注释的文法。

File: python/paddle/trainer_config_helpers/optimizers.py [hunk 12]
```diff
@@ -246,18 +268,29 @@ class DecayedAdaGradOptimizer(BaseSGDOptimizer):
     :type rho: float
     :param epsilon: The :math:`\\epsilon` parameter in that equation.
     :type epsilon: float
+    param momentum: add momentum for DecayedAdaGradOptimizer
+                    DecayedAdaGradOptimizer support add momentum property
```
Comment (reyoung) [inline python/paddle/trainer_config_helpers/optimizers.py]: 英文文法不对。请检查一下。

至少也是support => supports

不过，这里也没必要说为什么这么做，而是说这东西是什么就行了。


`param momentum:` => `:param momentum:` 前面少了个冒号。

**Review by reyoung — CHANGES_REQUESTED** [review]

File: python/paddle/trainer/config_parser.py (hunk 5 above)
Comment (backyes) [inline python/paddle/trainer/config_parser.py]: 注释解释了为什么，我觉得挺好的，也非常重要。call_once 非常直观明了，但是为什么倒不是一眼能看出来。

File: python/paddle/trainer/config_parser.py (hunk 2 above)
Comment (backyes) [inline python/paddle/trainer/config_parser.py]: 如果func函数带了一个is_called成员变量，那么这个装饰器就不能修饰这个func，所以这里check一下也比较合理。 另外显示的setattr也增加可读性。

File: python/paddle/trainer_config_helpers/optimizers.py (hunk 12 above)
Comment (backyes) [inline python/paddle/trainer_config_helpers/optimizers.py]: done

File: python/paddle/trainer_config_helpers/optimizers.py (hunk 8 above)
Comment (backyes) [inline python/paddle/trainer_config_helpers/optimizers.py]: sparse用来check的，仅仅做check。因为标题所列所有算法都支持momentum，但是如果使能了momentum，那么都不允许设置sparse。

File: python/paddle/trainer_config_helpers/optimizers.py (hunk 7 above)
Comment (backyes) [inline python/paddle/trainer_config_helpers/optimizers.py]: 为了统一接口吧。MomentumOptimizer带了一个sparse 参数，其他同理在允许配置momentum的情况下，也带上一个sparse选项吧。 各个optimizer面向用户的配置感觉越统一约好。

**Comment by luotao1** [comment]
感谢您给PaddlePaddle贡献代码。由于Paddle V1/V2版本已不再维护，相关代码也已从develop分支上删除，因此关闭您的PR，欢迎您向Paddle最新版-Fluid贡献代码。
Thanks for contributing to PaddlePaddle! Since V1/V2 will not be maintained anymore, and related codes have been deleted from develop branch as well, we close this PR. Welcome to contribute to Fluid——the latest version of PaddlePaddle.
